src/Datasource/EntityTrait.php

```diff
@@ -437,12 +437,16 @@ public function toArray()
         $result = [];
         foreach ($this->visibleProperties() as $property) {
             $value = $this->get($property);
-            if (is_array($value) && isset($value[0]) && $value[0] instanceof EntityInterface) {
+            if (is_array($value)) {
                 $result[$property] = [];
                 foreach ($value as $k => $entity) {
-                    $result[$property][$k] = $entity->toArray();
+                    if (method_exists($entity, 'toArray')) {
+                        $result[$property][$k] = $entity->toArray();
+                    } else {
+                        $result[$property][$k] = $entity;
+                    }
                 }
-            } elseif ($value instanceof EntityInterface) {
+            } elseif (method_exists($value, 'toArray')) {
```

Comment:
It isn't the memory I would be concerned about. It is the wall time difference of method_exists() over instanceof. This may be an unfounded concern as I have not done any profiling around the two options.